### ISSUE TITLE
32-bit api.startstop hanging on Linux

### ISSUE THREAD

**Issue by derekbruening** [DynamoRIO/dynamorio#2000]
32-bit api.startstop sometimes hangs on Linux.

Often, it churns up the CPU: attaching gdb I see the main thread at the
write syscall for the "all done" print, before it's set should_spin to
false.  The other threads are all running the IBL looking up their
sideline_func.  sleep(0) can be a nop for some libc implementations: is that it?

But, even with sched_yield instead of sleep, they spin.  3 or 4 are in sched_yield at any one time
but the others are all doing dispatch operations.

The loop is like this:

```
  dispatch: target = 0x0804924b
  Entry into F4(0x0804924b).0x4f3bd040 (shared)
  Exit from F21(0x08048b80).0x4f3cd0e9 (shared)   
   (block ends with syscall)
  Entry into do_syscall to execute a non-ignorable system call
  system call 158
  Exit from system call
  post syscall: sysnum=0x0000009e, result=0x00000000 (0)

  dispatch: target = 0xf77a3bee
  Entry into F11(0xf77a3bee).0x4f3bd0ec (shared)
  Exit from sourceless ibl: bb ret        
   (target 0xf76129ac in cache but not lookup table)

  dispatch: target = 0xf76129ac
  Entry into F2(0xf76129ac).0x4f3bd018 (shared)
  Exit from sourceless ibl: bb ret        
   (target 0x0804924b in cache but not lookup table)
```

So we have several problems:

1) The syscall is not inlined
2) The ibl targets are never put into the table

Is this restricted to 32-bit b/c of vsyscall making it statically harder to
know the syscall #?  Can these 10 threads going through dispatch really
starve out the main thread for 90s?

To solve #1 some options are:

A) Treat "call *gs:0x10" as the syscall like we do for WOW64
B) Convert "call *gs:0x10" to a direct call via -indcall2direct
C) Whenever we see "mov $imm,eax; call *gs:0x10", store $imm in dcontext

Of course though that would help this test, -code_api does this:

    options->indcall2direct = false;
    options->inline_ignored_syscalls = false;

Maybe we could inline a syscall if there is no syscall event handler or the
syscall filter ignores it?

For #2 we need to know why it's not being added:

  Exit from sourceless ibl: bb ret        
   (target 0xf76359ac in cache but not lookup table)
  fragment_add_ibl_target tag 0xf76359ac, branch 0, F0 

  dispatch: target = 0xf76359ac
  Entry into F2(0xf76359ac).0x47909018 (shared)

So it fails to find the fragment for the target.  It's b/c it only looks
for a trace.  It's the old issue of a dispatch-syscall permanently
interrupting a trace chain: we're supposed to mark the post-syscall point
as a trace head b/c the pre-syscall is a trace.

But with -disable_traces we still get the hang even though the loop is tighter:

```
  Exit from F10(0xf77b6be0).0x4a4cd181 (shared)   
   (block ends with syscall)
  Entry into do_syscall to execute a non-ignorable system call
  system call 158
  Exit from system call
  post syscall: sysnum=0x0000009e, result=0x00000000 (0)
  finished handling system call

  dispatch: target = 0xf77b6bee
  Entry into F11(0xf77b6bee).0x4a4cd19c (shared)
```

The bottom line is that exiting to dispatch in a loop is enough to cause a
major perf problem.

**Comment by derekbruening** [comment]
It seemed too extreme to just be starvation and it is.  While the perf issues with sysenter remain, the main thread made no progress because of a bug in the new vsyscall sequence handling in #1939.

Watching in gdb it is being scheduled but it seems to always be in dispatch_enter_native.  The log says:

```
Entry into native_exec after intercepted syscall
hit post-sysenter hook while native
```

Here's the bug:

```
            dcontext->native_exec_postsyscall = vsyscall_syscall_end_pc;
```

Pre-4.4.8, we put the displaced instructions at the end_pc, but not for
4.4.8+.  This causes it to never make progress as it keeps hitting the hook
after an int 0x80 that fails with ENOSYS over and over after an initial
failure:

```
       xax = 0x00000016
hit post-sysenter hook while native
        xax = 0xffffffff
hit post-sysenter hook while native
        xax = 0xffffffda
hit post-sysenter hook while native
        xax = 0xffffffda
```

**Comment by derekbruening** [comment]
I'm going to put the no-progress bug under #1939 and keep this open as it has a good description of the trace barrier issues for 32-bit Linux.
